File: tasks/expr-eval/javascript/environment/src/parse.js
"use strict";

const TOP_LEVEL = ["config", "programs"];

/** Only the top-level key set is checked so far. */
function parseDocument(value) {
  if (Object.keys(value).sort().join(",") !== TOP_LEVEL.join(",")) {
    throw new Error("malformed document");
  }
  return { maxDepth: value.config.maxDepth, programs: value.programs };
}

/** Decimal literals and the two operators that are understood so far. */
function tokenize(source) {
  const tokens = [];
  let index = 0;
  while (index < source.length) {
    const character = source[index];
    if (character === " ") {
      index += 1;
      continue;
    }
    if (character === "+" || character === "*") {
      tokens.push({ value: character, at: index });
      index += 1;
      continue;
    }
    const start = index;
    while (index < source.length && source[index] >= "0" && source[index] <= "9") {
      index += 1;
    }
    if (index === start) {
      throw new Error(`unexpected character at ${index}`);
    }
    tokens.push({ value: Number(source.slice(start, index)), at: start });
  }
  return tokens;
}

module.exports = { parseDocument, tokenize };

File: tasks/expr-eval/python/src/parse.py
"""Document handling and a first pass at reading program sources."""

TOP_LEVEL = ["config", "programs"]


def parse_document(value):
    """Only the top-level key set is checked so far."""
    if sorted(value.keys()) != TOP_LEVEL:
        raise ValueError("malformed document")
    return {
        "max_depth": value["config"]["maxDepth"],
        "programs": value["programs"],
    }


def tokenize(source):
    """Decimal literals and the two operators that are understood so far."""
    tokens = []
    index = 0
    while index < len(source):
        character = source[index]
        if character == " ":
            index += 1
            continue
        if character in "+*":
            tokens.append((character, index))
            index += 1
            continue
        start = index
        while index < len(source) and source[index].isdigit():
            index += 1
        if index == start:
            raise ValueError(f"unexpected character at {index}")
        tokens.append((int(source[start:index]), start))
    return tokens
